File: src/Player.cpp
#include "Combatant.h"

using namespace godot;

void Player::_register_methods() {
    register_method("_process", &Player::_process);
	register_method("_physics_process", &Player::_physics_process);
    register_method("_ready", &Player::_ready);
	register_method("_input", &Player::_input);
	register_method("process_input", &Player::process_input);
	register_method("process_health_meter", &Player::process_health_meter);
	register_method("update_health", &Player::update_health);
    register_property("Max HP", &Player::maxHP, 100);
    register_property("Max Fuel", &Player::maxFuel, 100);
    register_property("Recharge Rate", &Player::rechargeRate, 100.0f);
    register_property("Fuel Drain Rate", &Player::fuelDrainRate, 60.0f);
    register_property("Jet Force", &Player::jetForce, 3.0f);
    register_property("Normal Speed", &Player::normalSpeed, 7.0f);
    register_property("Sprint Speed", &Player::sprintSpeed, 14.0f);
    register_property("Overheal Decharge Rate", &Player::overhealDecrementRate, 0.25f);
    register_property("Aim Sensitivity", &Player::mouseSensitivity, 0.3f);
}

Player::Player() {

}

Player::~Player() {
    // add your cleanup here
}

void Player::_init() {
	_inputt = nullptr;
	_config = nullptr;
    _camera = nullptr;
    _gun = nullptr;
    _healthMeter = nullptr;
	_fuelMeter = nullptr;
    _healthNum = nullptr;
    _screenAni = nullptr;
    _boostSnd = nullptr, _medSnd = nullptr, _hurtSnd = nullptr, _landSnd = nullptr;
    alive = true;
    HPcounter = 0;
    fuel = 0;
    flameThrowerOn = false;
    overhealTimer = 0;
    canPlayBumpSound = false;
}
void Player::_ready() {
    _config = get_node<Config>("/root/Config");
    _boostSnd = get_node<AudioStreamPlayer>("boostSnd");
    _medSnd = get_node<AudioStreamPlayer>("medSnd");
    _hurtSnd = get_node<AudioStreamPlayer>("hurtSnd");
    _landSnd = get_node<AudioStreamPlayer>("landSnd");
    _head = get_node<Spatial>("Neck");
    _camera = get_node<Camera>("Neck/Camera");
    _gun = get_node<PlayerGun>("Neck/Gun");
    _healthMeter = get_node<TextureProgress>("HUD/HealthMeter");
    _fuelMeter = get_node<TextureProgress>("HUD/FuelMeter");
    _healthNum = get_node<Label>("HUD/HealthMeter/HealthNum");
    _screenAni = get_node<AnimationPlayer>("HUD/ScreenFlash/ScreenTransitions");
    _healthMeter->set_max(maxHP);
    _fuelMeter->set_max(maxFuel);
    mouseSensitivity = _config->mouseSensitivity;
    HP = maxHP;
    _healthNum->set_text(std::to_string(HPcounter).c_str());
    _healthMeter->set_value(HPcounter);
	_inputt = Input::get_singleton();
	_inputt->set_mouse_mode(_inputt->MOUSE_MODE_CAPTURED);

}
void Player::_process(const double p_delta) {
	//Analog stick Aiming
	_head->rotate_x(Math::deg2rad(_inputt->get_joy_axis(0, 3) * -mouseSensitivity));
	rotate_y(Math::deg2rad(-_inputt->get_joy_axis(0, 2)* mouseSensitivity));
	//overheal mechanics
	if (HP > _healthMeter->get_max())
	{
		//overheal = true;
		overhealTimer += p_delta;
		if (overhealTimer >= overhealDecrementRate)
		{
			overhealTimer = 0;
			HP--;
			_healthNum->set_text(std::to_string(HP).c_str());
		}
	}
	//else overheal = false;
}
void Player::_physics_process(const double p_delta)
{	
	//process_health_meter(p_delta);
	process_movement(p_delta); //it is very important that you do MoveAndSlide() before you do
	process_input(p_delta); //the IsOnFloor() calls that the input processing will do.
}
void Player::_input(Variant event) {
	mouseSensitivity = _config->mouseSensitivity;
	Input* _input = Input::get_singleton();
    if(_input == nullptr) return;   

	
	
	//Mouse Aiming
	Ref<InputEventMouseMotion> mouseEvent = event; //I'm not sure if this is how you're supposed to attempt a recast but it won't let me use Object::cast_to<>(). Using dynamic_cast doesn't work either.
	//Ref<InputEventMouseMotion> mouseEvent = Object::cast_to<InputEventMouseMotion>(event);
	//because of the above problem, I sincerely doubt that the null check below will actually work.
	if (mouseEvent != NULL && _input->get_mouse_mode() == _input->MOUSE_MODE_CAPTURED)
	{
		_head->rotate_x(Math::deg2rad(mouseEvent->get_relative().y * -mouseSensitivity));
		rotate_y(Math::deg2rad(-mouseEvent->get_relative().x * mouseSensitivity));
		//apply vertical clamping to rotation
		Vector3 cameraRot = _head->get_rotation_degrees();
		//TIL the GDnative C++ implementation of clamp() does not like int values under any circumstances. Must be real_t or float.
		cameraRot.x = Math::clamp(cameraRot.x, -85.0f, 85.0f);
		_head->set_rotation_degrees(cameraRot);
	}
}

void Player::process_input(const double p_delta){
	//  Walking
	
	Transform camXform = _camera->get_global_transform();
	Vector2 inputMovementVector = Vector2{0,0};
		
		
	dir = Vector3{0,0,0};
	if (_inputt->is_action_pressed("player_forward"))
		inputMovementVector.y += 1;
	if (_inputt->is_action_pressed("player_backward"))
		inputMovementVector.y -= 1;
	if (_inputt->is_action_pressed("player_left"))
		inputMovementVector.x -= 1;
	if (_inputt->is_action_pressed("player_right"))
		inputMovementVector.x += 1;

	inputMovementVector = inputMovementVector.normalized();
	
	if (alive)
	{
		//apply movement vector
		dir += -camXform.basis.z.normalized() * inputMovementVector.y;
		dir += camXform.basis.x.normalized() * inputMovementVector.x;

		//  Jumping / Jetpack thrust
		if (_inputt->is_action_pressed("player_jump") && _fuelMeter->get_value() > 0)
		{
			Godot::print("JUMP PLSS????");
			if (_inputt->is_action_just_pressed("player_jump")) _boostSnd->play();
		
			vel.y = (fuel / 10);
			fuel -= fuelDrainRate * p_delta;
		}

		//sprinting
		if (_inputt->is_action_pressed("player_sprint") && _fuelMeter->get_value() > 0)
		{
			if (_inputt->is_action_just_pressed("player_sprint")) _boostSnd->play();
			
			maxSpeed = sprintSpeed;
			fuel -= (fuelDrainRate / 3) * p_delta;
		}
		else maxSpeed = normalSpeed;

		//stop jetpack sounds
		if ((!_inputt->is_action_pressed("player_sprint") && !_inputt->is_action_pressed("player_jump")) || _fuelMeter->get_value() == 0) _boostSnd->stop();
		
		//recharging
		if (is_on_floor())
		{   
			if (canPlayBumpSound)
			{
				_landSnd->play(); //this doesn't quite work as predictably as I'd like. Oh well.
				canPlayBumpSound = false;
			}
			if (!_inputt->is_action_pressed("player_sprint") && !flameThrowerOn)
			{
				if (fuel < 100)
				{
					fuel += rechargeRate * p_delta;
				}
			}
		}
		else canPlayBumpSound = true;
	}
	else
	{
		get_tree()->change_scene("res://Scenes/GameOver.tscn");
	}
	fuel = Math::clamp(fuel, 0.0f, 100.0f);
	_fuelMeter->set_value(fuel);
}

void Player::process_health_meter(const double p_delta) {
    if (HP != HPcounter)
		{
			//flash the color between blue and white every frame
			if (_healthMeter->get_tint_progress() == Color{1,1,1,1})
			{
				_healthMeter->set_tint_progress(Color {0,0,1,1});
			}
			else _healthMeter->set_tint_progress(Color {1,1,1,1});
			//increment/decrement the meter towards the HP value every frame
			if (HP > HPcounter)
			{
				HPcounter++;
				
			}
			else if (HP < HPcounter)
			{
				HPcounter--;
			}
		}
		else
		{
			_healthMeter->set_tint_progress(Color {0,0,1,1});
		}
		_healthNum->set_text(std::to_string(HPcounter).c_str());
		_healthMeter->set_value(HPcounter);
}
void Player::update_health(const int delta) {
    if (delta < 0)
	{
		if (alive)
		{
			_hurtSnd->play();
			_screenAni->play("Hurt");
		}
	}
	else
	{
		fuel = 100;
		_medSnd->play();
	}
	HP += delta;
	if (HP <= 0)
	{
		alive = false;
		_gun->disabled = true;
	}
}

File: project/Scripts/Player.cs
using Godot;
using System;

public class Player : Combatant
{
	Config config;
	//PHYSICS VARIABLES--------------------------------
	[Export]
	public float RechargeRate = 100, FuelDrainRate = 60;
	[Export]
	public float JetForce = 3f;
	[Export]
	public float normalSpeed = 7f;
	[Export]
	public float sprintSpeed = 14f;
	[Export]
	public float mouseSensitivity = 0.3f;
	
	//GAMEPLAY VARIABLES-------------------------------------
	public bool alive = true;
	[Export]
	public int maxHP = 100; //in case I want to make max HP/max fuel upgrades a thing later.
	[Export]
	public int maxFuel = 100;
	int HPcounter = 0;
	public float fuel = 0;
	public bool flameThrowerOn = false;
	public float overhealDecrementRate = .25f;
	float overhealTimer = 0;
	bool needToPlayLandSound = false;

	//COMPONENT VARIABLES------------------------------------
	private Camera camera;
	public PlayerGun gun;
	private TextureProgress healthMeter, fuelMeter;
	private Label healthNum;
	private AnimationPlayer screenAni;
	protected AudioStreamPlayer boostSnd, medSnd, hurtSnd, landSnd;
	// Called when the node enters the scene tree for the first time.
	public override void _Ready()
	{
		config = GetNode<Config>("/root/Config");
		boostSnd = GetNode<AudioStreamPlayer>("boostSnd");
		medSnd = GetNode<AudioStreamPlayer>("medSnd");
		hurtSnd = GetNode<AudioStreamPlayer>("hurtSnd");
		landSnd = GetNode<AudioStreamPlayer>("landSnd");
		head = GetNode<Spatial>("Neck");
		camera = GetNode<Camera>("Neck/Camera");
		gun = GetNode<PlayerGun>("Neck/Gun");
		healthMeter = GetNode<TextureProgress>("HUD/HealthMeter");
		fuelMeter = GetNode<TextureProgress>("HUD/FuelMeter");
		healthNum = GetNode<Label>("HUD/HealthMeter/HealthNum");
		screenAni = GetNode<AnimationPlayer>("HUD/ScreenFlash/ScreenTransitions");
		Input.MouseMode = Input.MouseModeEnum.Captured;
		mouseSensitivity = config.mouseSensitivity;
		HP = maxHP;
		healthMeter.MaxValue = maxHP;
		fuelMeter.MaxValue = maxFuel;
		healthNum.Text = HPcounter.ToString();
		healthMeter.Value = HPcounter;
	}
    public override void _Process(float delta)
    {
		// Analog stick aiming
		head.RotateX(Mathf.Deg2Rad(Input.GetJoyAxis(0, 3) * -mouseSensitivity));
		RotateY(Mathf.Deg2Rad(-Input.GetJoyAxis(0, 2)* mouseSensitivity));
		//overheal mechanics
		if (HP > healthMeter.MaxValue)
		{
			//overheal = true;
			overhealTimer += delta;
			if (overhealTimer >= overhealDecrementRate)
			{
				overhealTimer = 0;
				HP--;
				healthNum.Text = HP.ToString();
			}
		}
		//else overheal = false;
    }
    
	public override void _PhysicsProcess(float delta)
	{
		ProcessHealthMeter(delta);
		ProcessMovement(delta); //it is very important that you do MoveAndSlide() before you do
		ProcessInput(delta); //the IsOnFloor() calls that the input processing will do.
	}

	private void ProcessInput(float delta)
	{
		//  Walking
		Transform camXform = camera.GlobalTransform;

		Vector2 inputMovementVector = new Vector2();
		
		
			dir = new Vector3();
			if (Input.IsActionPressed("player_forward"))
			inputMovementVector.y += 1;
			if (Input.IsActionPressed("player_backward"))
				inputMovementVector.y -= 1;
			if (Input.IsActionPressed("player_left"))
				inputMovementVector.x -= 1;
			if (Input.IsActionPressed("player_right"))
				inputMovementVector.x += 1;
		
		

		inputMovementVector = inputMovementVector.Normalized();

		if (alive)
		{
			//apply movement vector
			dir += -camXform.basis.z.Normalized() * inputMovementVector.y;
			dir += camXform.basis.x.Normalized() * inputMovementVector.x;
			
			
			//  Jumping / Jetpack thrust
			if (Input.IsActionPressed("player_jump") && fuelMeter.Value > 0)
			{
				if (Input.IsActionJustPressed("player_jump")) boostSnd.Play();
				
				vel.y = (fuel / 10);
				fuel -= FuelDrainRate * delta;
			}

			//sprinting
			if (Input.IsActionPressed("player_sprint") && fuelMeter.Value > 0)
			{
				if (Input.IsActionJustPressed("player_sprint")) boostSnd.Play();
				
				MaxSpeed = sprintSpeed;
				fuel -= (FuelDrainRate / 3) * delta;
			}
			else MaxSpeed = normalSpeed;

			//stop jetpack sounds
			if ((!Input.IsActionPressed("player_sprint") && !Input.IsActionPressed("player_jump")) || fuelMeter.Value == 0) boostSnd.Stop();
			
			//recharging
			if (IsOnFloor())
			{   
				if (needToPlayLandSound)
				{
					landSnd.Play(); //this doesn't quite work as predictably as I'd like. Oh well.
					needToPlayLandSound = false;
				}
				if (!Input.IsActionPressed("player_sprint") && !flameThrowerOn)
				{
					if (fuel < 100)
					{
						fuel += RechargeRate * delta;
					}
				}
			}
			else needToPlayLandSound = true;
		}
		else
		{
			GetTree().ChangeScene("res://Scenes/GameOver.tscn");
		}
		fuel = Mathf.Clamp(fuel, 0, 100);
		fuelMeter.Value = fuel;
	}
	public override void _Input(InputEvent @event)
	{
		mouseSensitivity = config.mouseSensitivity;
		if (@event is InputEventMouseMotion && Input.MouseMode == Input.MouseModeEnum.Captured)
		{
			InputEventMouseMotion mouseEvent = @event as InputEventMouseMotion;
			head.RotateX(Mathf.Deg2Rad(mouseEvent.Relative.y * -mouseSensitivity));
			RotateY(Mathf.Deg2Rad(-mouseEvent.Relative.x * mouseSensitivity));
			//apply vertical clamping to rotation
			Vector3 cameraRot = head.RotationDegrees;
			cameraRot.x = Mathf.Clamp(cameraRot.x, -85, 85);
			head.RotationDegrees = cameraRot;
		}
	}
	private void ProcessHealthMeter(float delta) //this is how I do health counter rolling
	{
		
		if (HP != HPcounter)
		{
			//flash the color between blue and white every frame
			if (healthMeter.TintProgress == Colors.White)
			{
				healthMeter.TintProgress = new Color(0,0,1,1);
			}
			else healthMeter.TintProgress = new Color(Colors.White);
			//increment/decrement the meter towards the HP value every frame
			if (HP > HPcounter)
			{
				HPcounter++;
				
			}
			else if (HP < HPcounter)
			{
				HPcounter--;
			}
		}
		else
		{
			healthMeter.TintProgress = new Color(0,0,1,1);
		}
		healthNum.Text = HPcounter.ToString();
		healthMeter.Value = HPcounter;
	}
	public override void UpdateHealth(int delta)
	{
		if (delta < 0)
		{
			if (alive)
			{
				hurtSnd.Play();
				screenAni.Play("Hurt");
			}
		}
		else
		{
			fuel = 100;
			medSnd.Play();
		}
		HP += delta;
		//healthMeter.Value = HP;
		//healthNum.Text = HP.ToString();
		if (HP <= 0)
		{
			alive = false;
			gun.disabled = true;
		}
	}
	
}
